### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by tekktrik**
Probably the most complex typing PR I've done, so definitely would appreciate a thorough look from someone else @___@  I had to learn a few things for this one, so hopefully I implemented it all correctly!

### CODE REVIEW

File: adafruit_requests.py [hunk 1]
```diff
@@ -37,22 +37,51 @@
...
+    InterfaceType = TypeVar("InterfaceType", ESP_SPIcontrol, WIZNET5K, FONA)
+
+except ImportError:
+    pass
 
 # CircuitPython 6.0 does not have the bytearray.split method.
 # This function emulates buf.split(needle)[0], which is the functionality
 # required.
-def _buffer_split0(buf, needle):
+def _buffer_split0(buf, needle: bytes):
```
Comment (dhalbert) [inline adafruit_requests.py:73]: I think `bytes` is unnecessarily narrow here. You could use `_typing.ReadableBuffer`. Or maybe it is only `Union[bytes, bytearray]` ?

File: adafruit_requests.py [hunk 2]
```diff
@@ -652,16 +710,18 @@ def connect(self, address):
...
+    def wrap_socket(
+        self, socket: SocketType, server_hostname: Optional[str] = None
+    ) -> _FakeSSLSocket:
         """Return the same socket"""
         # pylint: disable=unused-argument
         return _FakeSSLSocket(socket, self._iface.TLS_MODE)
 
 
-def set_socket(sock, iface=None):
+def set_socket(sock, iface: Optional[InterfaceType] = None) -> None:
```
Comment (dhalbert) [inline adafruit_requests.py:724]: What is the type of `sock`?

File: adafruit_requests.py [hunk 3]
```diff
@@ -672,7 +732,15 @@ def set_socket(sock, iface=None):
     sock.set_interface(iface)
 
 
-def request(method, url, data=None, json=None, headers=None, stream=False, timeout=1):
+def request(
+    method,
```
Comment (dhalbert) [inline adafruit_requests.py:736]: What is the type of `method`?

File: adafruit_requests.py [hunk 4]
```diff
@@ -187,7 +221,9 @@ def _readto(self, first, second=b""):
 
         return b""
 
-    def _read_from_buffer(self, buf=None, nbytes=None):
+    def _read_from_buffer(
+        self, buf: Optional[bytearray] = None, nbytes: Optional[int] = None
```
Comment (dhalbert) [inline adafruit_requests.py:225]: Could `buf` be something else besides a `bytearray` ?

**Review by dhalbert — CHANGES_REQUESTED** [review]
Thanks for this difficult annotation. Just a few q's.

File: adafruit_requests.py (hunk 1 above)
Comment (tekktrik) [reply adafruit_requests.py:73]: I've had issues with importing `_typing` if I remember correctly, but also `Union[bytes, bytearray]` seemed more like what it was intended for

File: adafruit_requests.py (hunk 2 above)
Comment (tekktrik) [reply adafruit_requests.py:724]: `sock` is actually a module, either CircuitPython's `socketpool` or CPython's `socket` module.  Basically, it's any module which implements the `socket` class.

File: adafruit_requests.py (hunk 3 above)
Comment (tekktrik) [reply adafruit_requests.py:736]: Also good catch!  I must missed this.  If I remember correctcly, `method` is a string defining the request method (GET, PUT, POST, etc)

File: adafruit_requests.py (hunk 4 above)
Comment (tekktrik) [reply adafruit_requests.py:225]: I think this has to be of type `bytearray` based on this line:

https://github.com/tekktrik/Adafruit_CircuitPython_Requests/blob/72a300f9e339f4c9cd2e2364517bbf9aa997474d/adafruit_requests.py#L234

**Comment by tekktrik** [comment]
Change # 1 - Changed to `Union[bytes, bytearray]` and also added the missing type hint for `buf`
Change # 2 - Added
Change # 3 - Added
Change # 4 - Left as `bytearray` for now but I'll change it if I misunderstood and that parameter can be of type `bytes`

**Review by dhalbert — APPROVED** [review]
Thanks for the iteration!

**Pull request merged**